### ISSUE TITLE
deepseek coder llm api

### ISSUE THREAD

**Issue by irthomasthomas** [irthomasthomas/undecidability#452]
- [ ] [deepseek coder llm api](deepseek.com)

Relational databases like PostgreSQL and operating system kernels like Linux are foundational building blocks for computing systems. Every "application" is, in essence, an "application" of those building blocks to some purpose: Twitter applies databases to write and read tweets; it applies an operating system to operate the hardware required to do so.

Large language models (LLMs) and other complex data type generative models are a new building block at this foundational layer. As an ML researcher, I have many useful intuitions about what can be built with this novel building block and how to improve it. However, the people who have spent years building with those other blocks have many more. Many of these insights are surprising to someone who has spent more time with neural networks than with the network stack.

This post is a collection of short explainers for papers that directly draw from systems metaphors in designing improvements for LLMs, ranging from novel approaches to pretraining to inference-time speed optimizations. Perhaps most interesting are *prompting strategies* that make LLMs act more like kernels.

It is intended to get researchers and hackers of LLMs interested in systems and vice versa. We'll proceed by first introducing an idea from the systems level – roughly, the interface of hardware, kernel, and application – and then showing how it has been applied to LLMs in recent research.

**Speculative Execution**

Speculative Execution in Processors

Sometimes bottlenecks are mitigated by executing the work that most likely follows them: we "speculatively execute" work that will probably, but not necessarily, come after our bottleneck.

For example, rather than waiting for the result of a conditional to be available in order to determine which branch to execute, the processor could simply proceed with the "most likely" branch, based on compile-time hints or runtime statistics. This particular technique for speculative execution is called "branch prediction".

This can lead to large speedups when a particular branch of a conditional is taken only rarely (e.g., unhappy/error-handling paths) and requires a lot of work to execute (e.g., is subject to cache misses).

Speculative execution is infamous thanks to the [Spectre vulnerability](https://spectreattack.com/), which manipulates runtime statistics to "trick" the processor into executing out-of-bounds memory reads.

Undaunted by that cautionary tale of a multi-billion-dollar vulnerability, let's consider how we might replicate this pattern of executing cheap but possibly wrong operations to speed up LLM inference.

**Speculative Execution in LLMs: Accelerating Large Language Model Decoding with Speculative Sampling and Fast Inference from Transformers via Speculative Decoding**

Two roughly-simultaneous papers from Google Research and DeepMind, perennial frenemies, demonstrated roughly-identical methods for large speedups in LLM generation by using cheaper models to predict certain tokens without changing the system's output probabilities. In essence, speculatively generating a number of likely output tokens before the actual LLM is used to compute the final outputs.

The intuition behind speculative sampling is that some of the tokens in the output are very easy to guess – a period will come at the end of this sentence, for example – and so can be predicted very cheaply.

These methods further take advantage of the fact that computing the logprobs for a prompt + K tokens can be done in parallel. That makes it much cheaper than sampling K tokens to follow a prompt, which must be done serially – a painful and unavoidable fact of autoregressive models like Transformers or RNNs.

The K tokens following the prompt are generated by some cheap strategy – a smaller language model or even a heuristic rule. We then compute the logprobs for the prompt + K tokens in the large model, in parallel, and then throw out only those tokens that are improbable according to the large model.

Importantly, we reject samples in direct proportion to how improbable the large model thinks they are, so the combined system has the exact same output behavior as the original large model, but generally runs faster. This is a classic trick, rejection sampling, that goes back to the very first (analog!) computers in the 1940s.

For temperature 0, aka deterministic sampling, this equates to throwing out all tokens that follow the first wrong token put out by the speculative process.

The basic flow is captured well in this animation of speculative sampling from an excellent explainer by Joao Gante of Hugging Face:

![Speculative Sampling](https://huggingface.co/datasets/huggingface/course-datasets/resolve/main/speculative_sampling.gif)

Update 2023-12-30: the topics discussed here are also covered in this [recorded talk from the Scale By The Bay 2023 conference](https://www.youtube.com/watch?v=your-video-id).

#### Suggested labels
- ['computing_systems']
